### PR TITLE
feat(graphql_semantic): build semantic model from AST

### PR DESCRIPTION
**Pull request by vohoanglong0107**
## Summary

Depends on https://github.com/biomejs/biome/pull/3427

Build a Semantic model for GraphQL from AST.

Unlike in JS, in GraphQL, a variable can be referenced in a fragment, which in turn can be referenced by multiple operations, and then defined in those operations. This results in a variable being potentially defined in multiple locations.

```rust
fragment HeroDetails on Character {
    name(startWith: $start)
}

query HeroWithH($start: String = "H") {
	...HeroDetails
}

query HeroWithO($start: String = "O") {
	...HeroDetails
}
```

In this case, the $start variable is defined in both the `HeroWithH` and `HeroWithO` variable definitions, and referenced at `(startWith: $start)` in `HeroDetails` fragment.

To address this, each operation and fragment will be associated with a scope. The scope will keep track of every referenced fragment, allowing it to track every referenced variable, either directly referenced or indirectly through a referenced fragment. Then, it will try to bind every referenced variable to its declaration inside an operation definition.

## Test Plan

All the calls to retrieve `binding` and `reference` should correctly return the respective bindings and references of the AST node.

### CODE REVIEW

**Comment by ematipico** [comment]
Are you able to separate the changes you made to the parser from the actual model changes? Or are they connected? It doesn't seem they are by reading the description of the PR

**Comment by vohoanglong0107** [comment]
Ah sure

File: crates/biome_graphql_semantic/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+
+[package]
+authors.workspace    = true
+categories.workspace = true
+description          = "<DESCRIPTION>"
```
Comment (ematipico) [inline crates/biome_graphql_semantic/Cargo.toml:5]: Update the description

File: crates/biome_graphql_semantic/src/semantic_model/builder.rs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            references: self.references,
+            references_by_start: self.references_by_start,
+            references_to_bindings: self.references_to_bindings,
+            unresolved_references: self.unresolved_references,
+            unresolved_variable_references: self.unresolved_variable_references,
+        };
+        SemanticModel::new(data)
+    }
+
+    fn is_builtin_type(name: &str) -> bool {
```
Comment (ematipico) [inline crates/biome_graphql_semantic/src/semantic_model/builder.rs:144]:
```suggestion
    const fn is_builtin_type(name: &str) -> bool {
```

File: crates/biome_graphql_semantic/src/semantic_model/builder.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        matches!(name, "String" | "Int" | "Float" | "Boolean" | "ID")
+    }
+
+    fn is_builtin_directive(name: &str) -> bool {
```
Comment (ematipico) [inline crates/biome_graphql_semantic/src/semantic_model/builder.rs:148]:
```suggestion
    const fn is_builtin_directive(name: &str) -> bool {
```

**Review by ematipico — APPROVED** [review]
This is great!

File: crates/biome_graphql_semantic/src/events.rs [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+
+/// A binding name is either a type or a value.
+///
+/// A value refer can refer to a directive or an operation.
+/// A type can refer to a type definition like a scalar or an object.
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+enum BindingName {
+    Type(TokenText),
+    Value(TokenText),
+}
```
[the remark this answers is not in the document]
Comment (vohoanglong0107) [reply crates/biome_graphql_semantic/src/events.rs:122]: Yeah, consider the following example:

```graphql
query Query {}

type Query

schema {
  query: Query
}
```

Here `Query` is used both as the name of the query, and the type of query in the schema

File: crates/biome_graphql_semantic/src/semantic_model/builder.rs (hunk 2 above)
Comment (vohoanglong0107) [reply crates/biome_graphql_semantic/src/semantic_model/builder.rs:144]: I think matching str in const fn is not yet supported by rust. https://github.com/rust-lang/rust/issues/90237

Comment (ematipico) [reply crates/biome_graphql_semantic/src/semantic_model/builder.rs:144]: ah I missed that, sorry

**Review by ematipico — APPROVED** [review]

**Pull request merged**